### PR TITLE
Search UI

### PR DESCRIPTION
**Pull request by alfredgrip**
This PR rewrites search in frontend so that it uses the Meilisearch endpoint introduced in #611 

It also adds a "advanced search" that allows filtering search on different indexes

It also fixes some minor TS misses that slipped through in the #611 PR

To try out this PR, please ensure that Meilisearch is up and running on your Docker (rerun the `dev/setup_db.sh` script) and then head to `localhost:.../admin/debug` and press `Sync with Meilisearch`

### CODE REVIEW

File: src/routes/(app)/search/+page.server.ts [hunk 1]
```diff
@@ -1,14 +1,45 @@
...
+
+    const url = new URL("/api/search", event.request.url);
+    url.searchParams.set("query", query);
+    url.searchParams.set("indexes", JSON.stringify(indexes));
+    url.searchParams.set("language", language);
+    const response = await event.fetch(url);
+    if (!response.ok) {
+      // silently fail
+      return;
+    }
```
Comment (Isak-Kallini) [inline src/routes/(app)/search/+page.server.ts:34]: Who no error handling? We should at least log errors so we can investigate any problems in the future

File: src/routes/(app)/search/+page.server.ts [hunk 2]
```diff
@@ -1,14 +1,45 @@
...
+    const includeEvents = data.get("events") === "on";
+    const includeArticles = data.get("articles") === "on";
+    const includePositions = data.get("positions") === "on";
+    const includeSongs = data.get("songs") === "on";
+    const indexes = [];
+    if (includeMembers) indexes.push("members");
+    if (includeEvents) indexes.push("events");
+    if (includeArticles) indexes.push("articles");
+    if (includePositions) indexes.push("positions");
+    if (includeSongs) indexes.push("songs");
```
Comment (Isak-Kallini) [inline src/routes/(app)/search/+page.server.ts:22]: Nitpick but this could be shorter, something like
```
const indexes = [
  "members",
  "events",
  "articles",
  "positions",
  "songs",
].filter((v) => data.get(v) === "on");
```

File: src/lib/components/search/ArticleSearchResult.svelte [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    <div class="avatar aspect-square w-8 overflow-hidden rounded-full">
+      <span class="i-mdi-newspaper text-2xl"></span>
+    </div>
+    <div>
+      <h4 class="text-balance">
+        {#if languageTag() === "en"}
+          {article.headerEn}
+        {:else}
+          {article.header}
+        {/if}
```
Comment (Isak-Kallini) [inline src/lib/components/search/ArticleSearchResult.svelte:21]: Could this be handled in the backend instead so we can avoid this kind of if else everywhere?

File: src/lib/components/search/GlobalSearch.svelte [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+    isOpen = true;
+    document.body.style.overflow = "hidden";
+  }
+
+  function close() {
+    dialog.close();
+    isOpen = false;
+    document.body.style.overflow = "auto";
+    // input = "";
+    // results = [];
```
Comment (Isak-Kallini) [inline src/lib/components/search/GlobalSearch.svelte:58]: Remove?

**Review by Isak-Kallini — CHANGES_REQUESTED** [review]
Search in english doesn't seem to be working, for example:
The bottom article comes up when searching for test
image
But when searching for its title it doesn't show up:
image
This is with language set to english, the article also comes up if i search for it's swedish title

Also if I press backspace while the input is empty the spinner appears until I type something

**Comment by alfredgrip** [comment]
> Search in english doesn't seem to be working

I found the issue for this. The `/search` page does a fetch to the api endpoint (which in turn makes a request to Meilisearch). When the server on the search page gets a request, the `event.locals.language` gets populated to the correct value by `hooks.server.ts`. But when the `/search` does a fetch to the API endpoint, the server there checks for `event.locals.language` but it defaults to swedish.
Tried some console logging and found this when trying to search with language set to `en`:
```
search param language is: en      // query parameter
search locals language is: en     // event.locals.language
API param language is: en         // query parameter
API locals language is: sv        // event.locals.language
```
So either the `hooks.server.ts` is faulty, or the `/search` server needs to do the fetch some other way. 

@danieladugyan Do you have any thoughts about this?

**Comment by alfredgrip** [comment]
> I found the issue for this. The `/search` page does a fetch to the api endpoint (which in turn makes a request to Meilisearch). When the server on the search page gets a request, the `event.locals.language` gets populated to the correct value by `hooks.server.ts`. But when the `/search` does a fetch to the API endpoint, the server there checks for `event.locals.language` but it defaults to swedish. Tried some console logging and found this when trying to search with language set to `en`:
> 
> ```
> search param language is: en      // query parameter
> search locals language is: en     // event.locals.language
> API param language is: en         // query parameter
> API locals language is: sv        // event.locals.language
> ```
> 
> So either the `hooks.server.ts` is faulty, or the `/search` server needs to do the fetch some other way. When the API page instead gets language from the query parameter, search in english works as intended.

I found that this can simply be fixed if the `/search` server does a request to `/en/api/search` if `locals.language == "en"`. I'm not a big fan of this, however. Having two "different" API endpoints that does the same, but which could easily be merged in to a single one by having a query parameter i.e. `?language="en"`, seems weird. I think it makes more sense to have a single one that accepts the language query parameter. What are your thoughts @Isak-Kallini @danieladugyan ?

**Comment by danieladugyan** [comment]
> I found that this can simply be fixed if the /search server does a request to /en/api/search if locals.language == "en". I'm not a big fan of this, however. Having two "different" API endpoints that does the same, but which could easily be merged in to a single one by having a query parameter i.e. ?language="en", seems weird. I think it makes more sense to have a single one that accepts the language query parameter. What are your thoughts @Isak-Kallini @danieladugyan ?

Sorry, that's my bad for introducing the bug. Anyways, that's exactly how I think it should work. It's consistent with how language is handled everywhere else in the app.
1. The request hits the server. 
2. `paraglide.js` looks at the URL and check if it has a language prefix, i.e `/en`, and sets the language in `event.locals` accordingly.
3. Route handlers use `event.locals.language` as needed - the Prisma `translationExtension` doing this automatically is enough in most cases.

In summary, if we're on the Swedish site, fetch normally, otherwise fetch with `'en'`  prepended. Ideally this would be done automatically by the library, but maybe we could make a `fetch` wrapper?

**Comment by alfredgrip** [comment]
I think it's weird though, specifics sent to an API endpoint should IMO be in the query parameters, not extracted from the URL. But I guess this works.

Just pushed some changes that fixes the english search and also handle translation mapping.


> but maybe we could make a fetch wrapper?

So we would write a wrapper for the already existing `event.fetch` wrapper from SvelteKit? 🫠

File: src/routes/(app)/search/+page.server.ts (hunk 1 above)
Comment (Isak-Kallini) [reply src/routes/(app)/search/+page.server.ts]: Could we either just log the error server side, or add error handling client side? Otherwise we just get the generic 500 internal server error which is not very nice

**Comment by Isak-Kallini** [comment]
I think that solution looks fine since, as Daniel says, that's how language is handled everywhere else.

File: src/routes/(app)/search/+page.server.ts (hunk 1 above)
Comment (alfredgrip) [reply src/routes/(app)/search/+page.server.ts]: Fixed in latest push!

**Comment by alfredgrip** [comment]
> I think that solution looks fine since, as Daniel says, that's how language is handled everywhere else. But should maybe use `i18n.resolveRoute('/api/search')`?

Didn't know that was a thing, but that's a nice way to handle it I guess!

File: src/lib/components/search/ArticleSearchResult.svelte [hunk 5]
```diff
@@ -0,0 +1,24 @@
+<script lang="ts">
+  import type { ArticleSearchReturnAttributes } from "$lib/search/searchTypes";
+  export let article: ArticleSearchReturnAttributes;
+</script>
+
+<li>
+  <a
+    href={"/news/" + article.slug}
+    class="border border-transparent focus:border-primary"
+    id=".search-result"
```
Comment (danieladugyan) [inline src/lib/components/search/ArticleSearchResult.svelte:10]: What does this id do? And why this it start with a period? AFAIK id's are supposed to be globally unique.

Comment (alfredgrip) [reply src/lib/components/search/ArticleSearchResult.svelte]: This is a part that enables keyboard navigation with the search results, used in `captureListItems()`
```ts
listItems = Array.from(formElement?.querySelectorAll("a")).filter(
  (a) => a.id === ".search-result",
);
```
I agree that this isn't idiomatic, just changed it in my latest commit
